network-libp2p/src/connection_pool/behaviour.rs

```diff
@@ -113,25 +120,63 @@ struct ConnectionState<T> {
     desired_connections: usize,
     /// The set of services that this peer requires.
     required_services: Services,
+    /// Waker used for the next poll
+    waker: Option<Waker>,
+}
+
+impl<T: Clone + Ord + Unpin> Future for ConnectionState<T> {
+    type Output = ();
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        self.waker.store_waker(cx);
+
+        while let Some((id, sleep)) = self.unban_timeout.as_mut() {
+            if sleep.poll_unpin(cx).is_ready() {
+                let id = id.clone();
+                self.unmark_banned(id);
+                self.unban_timeout = None;
+
+                // While we are unbanning, try to unban more peers if their deadline also have been reached.
+                // Otherwise schedule a sleep until for the first peer that can be unbanned.
+                while let Some((next_id, deadline)) = self.unban_deadlines.pop_front() {
+                    if deadline.elapsed() >= self.ban_time {
+                        self.unmark_banned(next_id);
+                    } else {
+                        self.unban_timeout = Some((next_id, Box::pin(sleep_until(deadline))));
+                        break;
+                    }
+                }
+            } else {
+                break;
+            }
+        }
+
```

Comment:
```suggestion

        if self.unban_timeout.is_none() {
            self.waker.store_waker(cx);
        }
```
Only when no `Sleep` has registered to a waker we need to be able to wake up ourselves.